engine/runtime/action-source/src/file/obj.rs

```diff
@@ -0,0 +1,925 @@
+    content: &Bytes,
+    params: &ObjReaderParam,
+    sender: Sender<(Port, IngestionMessage)>,
+) -> Result<(), SourceError> {
+    let obj_data = parse_obj_content(content)?;
+
+    let materials = if params.parse_materials && !obj_data.material_libs.is_empty() {
+        let mut all_materials = HashMap::new();
+        for mtl_lib in &obj_data.material_libs {
+            let mtl_path = resolve_material_path(ctx, storage_resolver.clone(), mtl_lib).await?;
+            if let Some(mtl_path) = mtl_path {
+                match parse_mtl(ctx, storage_resolver.clone(), &mtl_path).await {
+                    Ok(mats) => {
+                        all_materials.extend(mats);
+                    }
+                    Err(e) => {
+                        tracing::warn!("Failed to parse material file {}: {}", mtl_lib, e);
+                    }
+                }
+            }
+        }
+        all_materials
+    } else {
+        HashMap::new()
+    };
+
+    if params.merge_groups {
+        let geometry = create_geometry_from_faces(&obj_data, &obj_data.faces, params)?;
+        let mut attributes = IndexMap::new();
+
+        attributes.insert(
+            Attribute::new("source"),
+            AttributeValue::String("OBJ".to_string()),
+        );
+
+        if !obj_data.objects.is_empty() {
+            attributes.insert(
+                Attribute::new("objects"),
+                AttributeValue::Array(
+                    obj_data
+                        .objects
+                        .iter()
+                        .map(|o| AttributeValue::String(o.clone()))
+                        .collect(),
+                ),
+            );
+        }
+
+        if !obj_data.groups.is_empty() {
+            attributes.insert(
+                Attribute::new("groups"),
+                AttributeValue::Array(
+                    obj_data
+                        .groups
+                        .iter()
+                        .map(|g| AttributeValue::String(g.clone()))
+                        .collect(),
+                ),
+            );
+        }
+
+        if !materials.is_empty() {
+            let used_materials: Vec<_> = obj_data
+                .faces
+                .iter()
+                .filter_map(|f| f.material.as_ref())
+                .collect::<std::collections::HashSet<_>>()
+                .into_iter()
+                .cloned()
+                .collect();
+
+            if !used_materials.is_empty() {
+                attributes.insert(
+                    Attribute::new("materials"),
+                    AttributeValue::Array(
+                        used_materials
+                            .iter()
+                            .map(|m| AttributeValue::String(m.clone()))
+                            .collect(),
+                    ),
+                );
+            }
+        }
+
+        attributes.insert(
+            Attribute::new("face_count"),
```

Comment:
Attribute names are supposed to be in camelCase, could you update this accordingly?